### PR TITLE
feat: add limits for Accelerate

### LINKED ISSUE #5346: remove preview from Accelerate
Remove the preview tag from Accelerate

### PR DESCRIPTION
**Pull request by ankur-arch**
- add a migration section for DP users
- add new error surfaces for Accelerate

 closes #5346 and #5288

### CODE REVIEW

File: content/400-reference/200-api-reference/250-error-reference.mdx [hunk 1]
```diff
@@ -521,3 +519,19 @@ Interactive transaction error:
...
+
+**QueryTimeout**—The global timeout of Accelerate has been exceeded. You can find the limit [here](/data-platform/accelerate/limitations#query-timeout-limit).
+
+#### <inlinecode>P6005</inlinecode>
+
+**InvalidParameters**—The user supplied invalid parameters. Currently only relevant for transaction methods. For example, setting a timeout that is too high. You can find the limit [here](/data-platform/accelerate/limitations#interactive-transactions-query-timeout-limit).
+
+#### <inlinecode>P6009</inlinecode>
+
+**ResponseSizeLimitExceeded**—The global response size limit of Accelerate has been exceeded. You can find the limit [here](/data-platform/accelerate/limitations#response-size-limit).
```
Comment (nikolasburk) [inline content/400-reference/200-api-reference/250-error-reference.mdx]: how about a different formatting. the "titles" you've written in bold seem to be "code" as well (I assume so because of the PascalCasing, so I think they should be written in backticks as well). also I think it makes sense to include them in the section heading already?! 

```suggestion
#### <inlinecode>P6004</inlinecode> (<inlinecode>QueryTimeout</inlinecode>)

The global timeout of Accelerate has been exceeded. You can find the limit [here](/data-platform/accelerate/limitations#query-timeout-limit).

#### <inlinecode>P6005</inlinecode> (<inlinecode>InvalidParameters</inlinecode>)

The user supplied invalid parameters. Currently only relevant for transaction methods. For example, setting a timeout that is too high. You can find the limit [here](/data-platform/accelerate/limitations#interactive-transactions-query-timeout-limit).

#### <inlinecode>P6009</inlinecode> (<inlinecode>ResponseSizeLimitExceeded</inlinecode>)

The global response size limit of Accelerate has been exceeded. You can find the limit [here](/data-platform/accelerate/limitations#response-size-limit).
```

File: content/800-data-platform/100-accelerate/200-getting-started.mdx [hunk 2]
```diff
@@ -15,20 +15,21 @@ toc: true
 To get started with Accelerate, you will need the following:
 
 - A GitHub account.
-- A project that uses [Prisma Client](/concepts/components/prisma-client) `4.16.1` or higher. (If your project is using interactive transactions, you need to use `5.1.1` or higher.)
+- A project that uses [Prisma Client](/concepts/components/prisma-client) `4.16.1` or higher. If your project is using interactive transactions, you need to use `5.1.1` or higher.
+  > We **recommend** using the latest version of Prisma.
```
Comment (nikolasburk) [inline content/800-data-platform/100-accelerate/200-getting-started.mdx]: if you want to include the recommendation, let's just add it in parantheses. otherwise it creates a lot of visual noise IMO,

```suggestion
- A project that uses [Prisma Client](/concepts/components/prisma-client) `4.16.1` or higher. If your project is using interactive transactions, you need to use `5.1.1` or higher. (We always recommend using the latest version of Prisma.)
```

File: content/800-data-platform/100-accelerate/200-getting-started.mdx [hunk 3]
```diff
@@ -15,20 +15,21 @@ toc: true
...
+In order to enable Accelerate, you can log in to [Prisma Data Platform](https://pris.ly/pdp) and create a new Cloud Project. Follow the instructions in the UI to add Accelerate.
 
 For further information on each individual step, you can find additional documentation [here](/data-platform/cloud-projects/platform/projects#configure-accelerate).
 
 At the end of the setup process, you'll obtain a connection string that connects to Accelerate. This connection string also contains an API key that you need to use when configuring Prisma Client to use Accelerate.
 
 ## 2. Use Accelerate in your application
 
-To get started using Accelerate, ensure you’re using Prisma Client `4.16.1` or higher.
+To get started using Accelerate, we recommend using the latest version of Prisma.
```
Comment (nikolasburk) [inline content/800-data-platform/100-accelerate/200-getting-started.mdx]:
```suggestion
To get started using Accelerate, we recommend using the [latest version of Prisma ORM](https://github.com/prisma/prisma/releases/).
```

File: content/800-data-platform/100-accelerate/200-getting-started.mdx [hunk 4]
```diff
@@ -180,3 +185,9 @@ You should now see improved performance for your cached queries.
 For information about which strategy best serves your application, see [Select a cache strategy](/data-platform/accelerate/concepts#selecting-a-cache-strategy).
 
 </Admonition>
+
+## Migrating from Data Proxy
+
+Similar to Data Proxy, Accelerate can be used as a **stand-alone connection pooler** for your application.
+
+To migrate from Data Proxy to Accelerate, make sure that you fulfill the [prerequisites](/data-platform/accelerate/getting-started#prerequisites) to use Accelerate in your application. After you have successfully [enable](/data-platform/accelerate/getting-started#1-enable-accelerate-in-a-cloud-project) Accelerate and generate an Accelerate connection string, follow the steps outlined [here](/data-platform/accelerate/getting-started#21-update-your-database-connection-string) to use Accelerate in your application.
```
Comment (nikolasburk) [inline content/800-data-platform/100-accelerate/200-getting-started.mdx]:
```suggestion
To migrate from Data Proxy to Accelerate, make sure that you fulfil the [prerequisites](/data-platform/accelerate/getting-started#prerequisites) to use Accelerate in your application. 

After you have successfully [enabled](/data-platform/accelerate/getting-started#1-enable-accelerate-in-a-cloud-project) Accelerate and generated an Accelerate connection string, follow the steps outlined [here](/data-platform/accelerate/getting-started#21-update-your-database-connection-string) to use Accelerate in your application.
```

File: content/800-data-platform/100-accelerate/200-getting-started.mdx [hunk 5]
```diff
@@ -180,3 +185,9 @@ You should now see improved performance for your cached queries.
 For information about which strategy best serves your application, see [Select a cache strategy](/data-platform/accelerate/concepts#selecting-a-cache-strategy).
 
 </Admonition>
+
+## Migrating from Data Proxy
+
+Similar to Data Proxy, Accelerate can be used as a **stand-alone connection pooler** for your application.
```
Comment (nikolasburk) [inline content/800-data-platform/100-accelerate/200-getting-started.mdx]: make this even more explicit:

```suggestion
Similar to Data Proxy, Accelerate can be used as a _stand-alone connection pooler_ for your application (Accelerate's caching feature is opt-in, so unless you explicitly specify a caching strategy for your Prisma Client queries, Accelerate will only serve as a connection pool for your application).
```

**Review by nikolasburk — CHANGES_REQUESTED** [review]
Great work Ankur! I've left a few comments but generally this looks good to me 💯 

One thing that I was surprised about though: The "migration from data proxy" only refers to the regular "Getting started"-section. Are we not planning to include a step-by-step migration guide specifically for Data Proxy users?

**Comment by ankur-arch** [comment]
> Are we not planning to include a step-by-step migration guide specifically for Data Proxy users?

@nikolasburk I'm open to making the steps more explicit.

File: content/800-data-platform/100-accelerate/200-getting-started.mdx [hunk 6]
```diff
@@ -72,7 +78,9 @@ datasource db {
 }
 ```
 
-Prisma Migrate and Prisma Studio will use the `directUrl` connection string rather than the one defined in `url` when this configuration is provided.
+Migrations and introspections will use the `directUrl` connection string rather than the one defined in `url` when this configuration is provided.
+
+> `directUrl` is useful for you to carry out migrations and introspections. However, you don't need `directUrl` to use Accelerate in your application.
```
Comment (petradonka) [inline content/800-data-platform/100-accelerate/200-getting-started.mdx:83]: Nice

File: content/800-data-platform/100-accelerate/200-getting-started.mdx [hunk 7]
```diff
@@ -82,26 +90,22 @@ Run the following command to install the Accelerate extension for Prisma Client:
...
-}
-```
-
 ### 2.3. Generate Prisma Client for Accelerate
 
 Run the following command to generate Prisma Client for Accelerate:
 
 ```terminal
-npx prisma generate --accelerate
+npx prisma generate --no-engine
```
Comment (petradonka) [inline content/800-data-platform/100-accelerate/200-getting-started.mdx:106]: This is only necessary if you want to strip out the engines, right? It's recommended, but not required? Is that right, and if so, should we mention that?

Comment (ankur-arch) [reply content/800-data-platform/100-accelerate/200-getting-started.mdx:106]: @petradonka okay, I'm explicitly mentioning that 🙏

Comment (petradonka) [reply content/800-data-platform/100-accelerate/200-getting-started.mdx:106]: Please only do if what I'm saying is actually correct 😄

**Review by nikolasburk — APPROVED** [review]
thanks for incorporating my suggestions, this looks good from my end now 👍

File: content/800-data-platform/100-accelerate/200-getting-started.mdx (hunk 7 above)
Comment (ankur-arch) [reply content/800-data-platform/100-accelerate/200-getting-started.mdx:106]: @petradonka - you are correct 😄 . I skipped the explanation initially because the section was as lengthy as it is. Here's the [reference](https://github.com/prisma/prisma/releases/tag/5.2.0).

I've updated the section in the latest commit.

Comment (petradonka) [reply content/800-data-platform/100-accelerate/200-getting-started.mdx:106]: @ankur-arch, should we mention what you should do if you're _not_ in serverless/edge and _not_ below 5.2.0? I believe the "happy path" is only implied right now?

Comment (ankur-arch) [reply content/800-data-platform/100-accelerate/200-getting-started.mdx:106]: > should we mention what you should do if you're *not* in serverless/edge and not below 5.2.0?

@petradonka I've included a small section on this at the beginning of the section (screenshot below). Imho that should cover all cases. I'm open to making edits if you have any suggestions on how I could make it more explicit 😄 .

image

File: content/400-reference/200-api-reference/250-error-reference.mdx [hunk 8]
```diff
@@ -384,7 +384,7 @@ The Schema Engine was previously called Migration Engine. This change was introd
 "Prisma Migrate could not create the shadow database. Please make sure the database user has permission to create databases. Read more about the shadow database (and workarounds) at [https://pris.ly/d/migrate-shadow](https://pris.ly/d/migrate-shadow).
 
 Original error: {error_code}<br />
-{inner_error}"
+"{inner_error}"
```
Comment (janpio) [inline content/400-reference/200-api-reference/250-error-reference.mdx]: This is incorrect. The `"` wraps the complete error message here.

File: content/400-reference/200-api-reference/250-error-reference.mdx [hunk 9]
```diff
@@ -432,8 +432,6 @@ Original error: {error_code}<br />
 
 "The schema of the introspected database was inconsistent: {explanation}"
 
-<!-- Error codes generated by https://github.com/mhwelander/doc-tools/blob/master/error_code_parser.py -->
```
Comment (janpio) [inline content/400-reference/200-api-reference/250-error-reference.mdx]: This is important (although unused right now)

File: content/400-reference/200-api-reference/250-error-reference.mdx [hunk 10]
```diff
@@ -521,3 +519,19 @@ Interactive transaction error:
 - Could not parse interactive transaction ID
 - Could not find Query Engine for the specified host and transaction ID
 - Could not start interactive transaction
+
+### Accelerate
+
+As Accelerate connections operate differently from local database connections, error codes are different from regular ones. In general, all Accelerate-related errors start with `P6xxx`.
```
Comment (janpio) [inline content/400-reference/200-api-reference/250-error-reference.mdx]: They are not different as far as I know, there are only some additional ones that are triggered by Accelerate only

File: content/800-data-platform/100-accelerate/200-getting-started.mdx [hunk 11]
```diff
@@ -55,7 +55,13 @@ DATABASE_URL="prisma://accelerate.prisma-data.net/?api_key=__API_KEY__"
 # DATABASE_URL="postgresql://user:password@host:port/db_name?schema=public"
 ```
 
-Prisma Migrate and Prisma Studio do not work with a `prisma://` connection string. In order to continue using these features add a new variable to the `.env` file named `DIRECT_DATABASE_URL` whose value is the direct database connection string:
+Prisma Migrate does not work with a `prisma://` connection string. In order to continue using these features add a new variable to the `.env` file named `DIRECT_DATABASE_URL` whose value is the direct database connection string:
```
Comment (janpio) [inline content/800-data-platform/100-accelerate/200-getting-started.mdx]:
```suggestion
Prisma Migrate and Introspection do not work with a `prisma://` connection string. In order to continue using these features add a new variable to the `.env` file named `DIRECT_DATABASE_URL` whose value is the direct database connection string:
```

File: content/800-data-platform/100-accelerate/200-getting-started.mdx [hunk 12]
```diff
@@ -82,26 +90,36 @@ Run the following command to install the Accelerate extension for Prisma Client:
...
+When using Prisma Accelerate in a Serverless or an Edge application, we recommend you to run the following command to generate Prisma Client:
 
 ```terminal
-npx prisma generate --accelerate
+npx prisma generate --no-engine
 ```
 
+<Admonition type="info">
+
+The `--no-engine` flag prevents a Query Engine file from being included in the generated Prisma Client, this ensures the bundle size of your application remains small, which helps improve performance in your Serverless and Edge applications.
```
Comment (janpio) [inline content/800-data-platform/100-accelerate/200-getting-started.mdx]: We do not have any measurements that show it influences the performance in any significant way.
It does make the bundle smaller though, which means you have more space for other stuff before you get an error that your archive is too big.

File: content/800-data-platform/100-accelerate/500-limitations.mdx [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+Accelerate has a global timeout of `10s` for each query. Reach out to [support@prisma.io](mailto:support@prisma.io) with your use case if your application requires a greater timeout value.
+
+## Interactive transactions query timeout limit
+
+Accelerate has a global timeout of `15s` for each [interactive transaction](/guides/performance-and-optimization/prisma-client-transactions-guide#interactive-transactions). Reach out to [support@prisma.io](mailto:support@prisma.io) with your use case if your application requires a greater timeout value.
+
+## Response size limit
+
+Accelerate has a global response size limit of `5MB`. Reach out to [support@prisma.io](mailto:support@prisma.io) with your use case if your application requires a larger response size.
```
Comment (janpio) [inline content/800-data-platform/100-accelerate/500-limitations.mdx:25]: These feel more important than the raw query and fluent API limitations. I suggest to move them up.

File: content/800-data-platform/100-accelerate/index.mdx [hunk 14]
```diff
@@ -1,7 +1,6 @@
 ---
 title: 'Accelerate'
 staticLink: true
-preview: true
```
Comment (janpio) [inline content/800-data-platform/100-accelerate/index.mdx:4]: Unrelated change?

Comment (ankur-arch) [reply content/800-data-platform/100-accelerate/index.mdx:4]: @janpio yeah, as it's a simple change, I did not want to create an additional PR for it.

File: content/800-data-platform/100-accelerate/200-getting-started.mdx (hunk 12 above)
Comment (ankur-arch) [reply content/800-data-platform/100-accelerate/200-getting-started.mdx]: > which helps improve performance in your Serverless and Edge applications.

removing the line above!

File: content/400-reference/200-api-reference/250-error-reference.mdx (hunk 10 above)
Comment (ankur-arch) [reply content/400-reference/200-api-reference/250-error-reference.mdx]: Updating copy to

> Accelerate-related errors start with `P6xxx`.

**Review by janpio — APPROVED** [review]
